File: src/main/java/tk/hiddenname/smarthome/service/task/impl/listener/ListenerFactory.java
package tk.hiddenname.smarthome.service.task.impl.listener;

import lombok.AllArgsConstructor;
import org.springframework.context.ApplicationContext;
import org.springframework.stereotype.Component;
import tk.hiddenname.smarthome.exception.NoSuchListenerException;
import tk.hiddenname.smarthome.exception.UnsupportedTriggerObjectTypeException;
import tk.hiddenname.smarthome.model.task.trigger.objects.TriggerObject;
import tk.hiddenname.smarthome.service.task.impl.listener.impl.ChangeDigitalSignalListener;

import javax.validation.constraints.NotNull;

@Component
@AllArgsConstructor
public class ListenerFactory {

    private final ApplicationContext ctx;

    public Listener create(@NotNull TriggerObject object, EventListener eventListener) throws NoSuchListenerException, UnsupportedTriggerObjectTypeException {
        Listener listener;

        switch (object.getAction()) {
            case CHANGE_DIGITAL_SIGNAL:
                listener = ctx.getBean(ChangeDigitalSignalListener.class, eventListener);
                break;
            default:
                throw new NoSuchListenerException();
        }

        listener.register(object);
        return listener;
    }
}

File: src/main/java/tk/hiddenname/smarthome/service/task/impl/listener/ListenerFactory.kt
package tk.hiddenname.smarthome.service.task.impl.listener

import org.springframework.context.ApplicationContext
import org.springframework.stereotype.Component
import tk.hiddenname.smarthome.exception.NoSuchListenerException
import tk.hiddenname.smarthome.exception.UnsupportedTriggerObjectTypeException
import tk.hiddenname.smarthome.model.task.trigger.TriggerAction
import tk.hiddenname.smarthome.model.task.trigger.objects.TriggerObject
import tk.hiddenname.smarthome.service.task.impl.listener.impl.ChangeDigitalSignalListener

@Component
class ListenerFactory(private val ctx: ApplicationContext) {

    @Throws(NoSuchListenerException::class, UnsupportedTriggerObjectTypeException::class)
    fun create(triggerObject: TriggerObject, eventListener: EventListener): Listener {
        val listener = when (triggerObject.action) {
            TriggerAction.CHANGE_DIGITAL_SIGNAL -> ctx.getBean(ChangeDigitalSignalListener::class.java, eventListener)
            else -> throw NoSuchListenerException()
        }
        listener.register(triggerObject)
        return listener
    }
}